### PR TITLE
Added music composition mutations and tests

### PR DESCRIPTION
**Pull request by pc2752**

### CODE REVIEW

File: tests/data/musiccomposition/EXPECTED_MUSICCOMPOSITION_ADD_EXAMPLEOF_COMPOSITION.txt [hunk 1]
```diff
@@ -0,0 +1,13 @@
+mutation {
+  AddDMusicCompositionExampleOfWork(
```
Comment (alastair) [inline tests/data/musiccomposition/EXPECTED_MUSICCOMPOSITION_ADD_EXAMPLEOF_COMPOSITION.txt]: Is this right?

File: trompace/mutations/musiccomposition.py [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+) {{
+    from {{
+      identifier
+    }}
+    to {{
+      identifier
+    }}
+}}'''
+
+MERGE_MUSIC_COMPOSITION_WORK_EXAMPLE_COMPOSITION = '''MergeMusicCompositionExampleOfWork(
```
Comment (alastair) [inline trompace/mutations/musiccomposition.py]: I believe that the behaviour here is that `Merge*` methods will add it if it doesn't exist, and do nothing if it does exist. If this is the case, we don't need the `Add*` version. Can you test this on the CE and remove Add if so?

File: trompace/mutations/musiccomposition.py [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+DELETE_MUSIC_COMPOSITION = '''DeleteMusicComposition(
+    {parameters}
+  ) {{
+    identifier
+  }}'''
+
+ADD_MUSIC_COMPOSITION_TO_CONTROL_ACTION_MUTATION = """AddActionInterfaceThingInterface (
+            from: {{identifier: "{identifier_2}", type: ControlAction}}
+            to: {{identifier: "{identifier_1}", type: MusicComposition}}
+            field: result
```
Comment (alastair) [inline trompace/mutations/musiccomposition.py]: I don't think this style of mutation exists any more. See #5

File: trompace/mutations/musiccomposition.py [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+                           language: str, inLanguage:str, formatin:str="text/html", name: str=None, description: str=None):
+    """Returns a mutation for creating a music composition object
+    Arguments:
+        title: The title of the page from which the music composition information was extracted.      
+        contributor: A person, an organization, or a service responsible for contributing the music composition to the web resource. This can be either a name or a base URL.
+        creator: The person, organization or service who created the thing the web resource is about.
+        source: The URL of the web resource to be represented by the node.
+        language: The language the metadata is written in. Currently supported languages are en,es,ca,nl,de,fr
+        inLanguage: The language of the music composition. Currently supported languages are en,es,ca,nl,de,fr
+        formatin: A MimeType of the format of the music composition, default is "text/html"
```
Comment (alastair) [inline trompace/mutations/musiccomposition.py]: This is not quite right. It should be the format of the _page_ that describes this composition. 
Why is this called formatin, and not format?

**Comment by alastair** [comment]
This PR can delete work.py too, if this code is duplicated

File: trompace/mutations/musiccomposition.py (hunk 4 above)
Comment (pc2752) [reply trompace/mutations/musiccomposition.py]: "format" is a reserved word for a built-in python function

Comment (alastair) [reply trompace/mutations/musiccomposition.py]: good point. To clarify this though, format is not a reserved keyword (like 'in' or 'class' or 'return'), it's just a function that exists in the default namespace. It wouldn't be a problem to use it (e.g. I often use 'id' as a field even though it's an existing function).
Looking at some guidelines, it looks like the normal suggestion with these variables is to add a `_` after the name: https://softwareengineering.stackexchange.com/a/339309
I recommend that we do the same

File: tests/data/musiccomposition/EXPECTED_MUSICCOMPOSITION.txt [hunk 5]
```diff
@@ -0,0 +1,13 @@
+mutation {
+  CreateMusicComposition(
+        title: "A Document"
+        contributor: "https://www.cpdl.org"
+        creator: "https://www.cpdl.org/A_Document"
+        format: "text/html"
+        language: en
+        inLanguage: en
```
[the remark this answers is not in the document]
Comment (pc2752) [reply tests/data/musiccomposition/EXPECTED_MUSICCOMPOSITION.txt:8]: Subject is also also required by the API, its not there in the modelling discussion, but I've added it for now.

**Comment by pc2752** [comment]
I'm not sure why the conflicts are not showing up on my end again, but I've made the changes required.

**Comment by pc2752** [comment]
git fetch origin and git pull just show "Already up-to-date", with no conflicts. But this page shows conflicts in 5 files.

**Comment by alastair** [comment]
what branch are you on when you're running those commands? your `add_composition` one? You'll have to merge master into the branch to fix these conflicts. Try this:

```
git checkout master
git fetch origin
git merge origin/master
git checkout add_composition
git merge master
```

**Comment by alastair** [comment]
great, thanks. Will get this merged today

**Comment by alastair** [comment]
I've made some changes to this, please take a look at them:

 * Remove work.py and test_composition.py, these were unneeded now that we have the musiccomposition version
 * I updated the structure of mutation_update_music_composition to not duplicate as much code

Can you please look at the tests for musiccomposition and update the following things:
 * Make the test files look more "correct" - e.g. EXPECTED_MUSICCOMPOSITION.txt seems to have field values that look more like a Document or a Person. These should at least look like an example from our modelling doc
 * rename the expected text files to lower-case, and a better name structure, like I did with person
 * Add some tests for create_musiccomposition and update_musiccomposition to test that exceptions are raised with bad data (e.g an invalid language or a format without a / in it)
 * add tests to check that optional values are not added if they're not given - e.g. do 2 tests for update_musiccomposition - one with only the required fields, check that only those fields are listed in the mutation, and one with all fields set, and make sure that all those fields are written in the mutation text.

thanks!

**Comment by alastair** [comment]
Looks good, but I see 3 failing tests:
<tests.test_musiccomposition.TestMusicComposition testMethod=test_add_broad_match>
<tests.test_musiccomposition.TestMusicComposition testMethod=test_update_all>
<tests.test_musiccomposition.TestMusicComposition testMethod=test_update_name>

looks like they're all to do with loading the expected text file

**Pull request merged**